sonic_y_cable/y_cable_base.py

```diff
@@ -0,0 +1,577 @@
+        TOR B. This means if the Y cable is actively routing, the "check_active_linked_tor_side"
+        API will now return Tor B. It also implies that if the link is actively routing on this port, Y cable
+        MUX will start forwarding packets from TOR B to NIC, and drop packets from TOR A to NIC
+        regardless of previous forwarding state.
+
+        Args:
+            port:
+                an Integer, the actual physical port connected to Y end of a Y cable which can toggle the MUX
+
+        Returns:
+            a Boolean, true if the toggle succeeded and false if it did not succeed.
+        """
+        raise NotImplementedError
+
+    def check_read_side(self, port):
+        """
+        This API specifically checks which side of the Y cable the reads are actually getting performed
+        from, either TOR A or TOR B or NIC and returns the value.
+
+        Args:
+            port:
+                an Integer, the actual physical port connected to Y end of a Y cable which can which side reading the MUX from
+
+        Returns:
+            an Integer, 1 if reading the Y cable from TOR A side(TOR 1).
+                      , 2 if reading the Y cable from TOR B side(TOR 2).
+                      , 0 if reading the Y cable from NIC side.
+                      , -1 if reading the Y cable API fails.
+        """
+
+        raise NotImplementedError
+
+    def check_mux_direction(self, port):
+        """
+        This API specifically checks which side of the Y cable mux is currently point to
+        and returns either TOR A or TOR B. Note that this API should return mux-direction
+        regardless of whether the link is active/routing state or not.
+
+        Args:
+            port:
+                an Integer, the actual physical port connected to a Y cable
+
+        Returns:
+            an Integer, 1 if the mux is pointing to TOR A .
+                      , 2 if the mux is pointing to TOR B.
+                      , -1 if checking which side mux is pointing to API fails.
+        """
+
+        raise NotImplementedError
+
+    def check_active_linked_tor_side(self, port):
+        """
+        This API specifically checks which side of the Y cable is actively linked and routing
+        and returns either TOR A or TOR B.
+
+        Args:
+            port:
+                an Integer, the actual physical port connected to a Y cable
+
+        Returns:
+            an Integer, 1 if TOR A is actively linked and routing(TOR 1).
+                      , 2 if TOR B is actively linked and routing(TOR 2).
+                      , 0 if nothing linked and actively routing
+                      , -1 if checking which side linked for routing API fails.
+        """
+
+        raise NotImplementedError
+
+    def check_if_link_is_active_for_NIC(self, port):
+        """
+        This API specifically checks if NIC side of the Y cable's link is active
```

Comment:
What is the definition we are using for 'active'? Does that mean the cable is plugged in, traffic is flowing through this link, etc.

Reply:
yes basically link is up means cable is up and configured correctly, show int status should show "oper" up

Reply:
Is this definition commonly known? It might be worth including this info in the docstring.

Reply:
Hmm maybe I can elaborate a little. I think "oper" up on Cli show interfaces status might confuse the vendors, but objectively  this API it's just their capability to check if all the parameters for transceiver initialization happened correctly, and the link is active. Actually when the lanes come up they send some packets internally on Serdes to check if lanes are functioning properly and a bunch of other things to check if link is healthy. While this generally translates to link is up, they cant tell if the link will be on SONiC side. So I think we can make this definition more concrete but maybe we should not put "show int status" output a cross check for them. Thoughts ?

Reply:
I think as long as the vendors know what this means, it's fine as is. I'll defer to your judgement here.